File: Part1_강의자료/Part1_연습문제_풀이/2473_세용액/2473_two_pointer.cpp
#include <algorithm>
#include <stdio.h>
#include <math.h>
#include <vector>
using namespace std;

int main()
{
    int N;
    scanf("%d", &N);

    vector<int> vec(N);
    for (int i = 0; i < N; i++)
        scanf("%d", &vec[i]);
    sort(vec.begin(), vec.end());

    int ans1 = vec[0];
    int ans2 = vec[1];
    int ans3 = vec[2];
    long long ans_abs = abs((long long)vec[0] + vec[1] + vec[2]);
    for (int i = 0; i < N - 2; i++) {
        int left_idx = i + 1;
        int right_idx = N - 1;
        while (left_idx < right_idx) {
            long long cur_sum = (long long)vec[i] + vec[left_idx] + vec[right_idx];
            long long cur_abs = abs(cur_sum);
            if (ans_abs > cur_abs) {
                ans1 = vec[i];
                ans2 = vec[left_idx];
                ans3 = vec[right_idx];
                ans_abs = cur_abs;
            }
            if (cur_sum <= 0)
                left_idx++;
            else
                right_idx--;
        }
    }

    printf("%d %d %d", ans1, ans2, ans3);
    return 0;
}

File: Part1_강의자료/Part1_연습문제_풀이/2473_세용액/2473.two_pointer.py
N = int(input())
arr = sorted(list(map(int,input().split())))
ans = arr[:3]
ans_abs = abs(sum(ans))

for i in range(N - 2):
    left_idx, right_idx = i + 1, N - 1
    while left_idx < right_idx:
        cur_sum = arr[i] + arr[left_idx] + arr[right_idx]
        cur_abs = abs(cur_sum)
        if ans_abs > cur_abs:
            ans = [arr[i], arr[left_idx], arr[right_idx]]
            ans_abs = cur_abs
        if cur_sum <= 0:
            left_idx += 1
        else:
            right_idx -= 1

print(' '.join(map(str, ans)))
